[Canary] Standalone build on monorepo is missing node_modules and messes up files

### Link to the code that reproduces this issue

https://github.com/PAXANDDOS/nextjs-canary-issue

### To Reproduce

1. Install dependencies with `bun i` (or equivalent)
2. Run `bun run build` (or equivalent) to build the standalone app
3. Open `.next/standalone` and see that `node_modules` is missing, and all application files are on the top level
4. Run `cp .next/static .next/standalone/.next/static`
5. Run `cd ./apps/web/.next/standalone && bun ./server.js` (or equivalent)
6. See the error in the console

**To Reproduce Working Version**
1. Open `apps/web/package.json` and change `next` version to `15.4.2-canary.21`
2. Run `bun update` (or equivalent)
3. Run `bun run build` (or equivalent) to build the standalone app
4. Run `cp .next/static .next/standalone/apps/web/.next/static`
5. Run `cd ./apps/web/.next/standalone && bun apps/web/server.js` (or equivalent)
6. See that the application works

### Current vs. Expected behavior

**Current behaviour**

- `node_modules` is completely missing from the standalone folder
- Application files (like `package.json` and `server.js`) are brought to the top level of the standalone folder, instead of being nested in `apps/web`
- After the standalone application is built and prepared, trying to start it would result in the next error:
```sh
error: Cannot find module './cpu-profile' from 'R:\issue\apps\web\node_modules\next\dist\server\lib\start-server.js'
```

**Expected behaviour**

- `node_modules` is at the top level of the standalone folder
- Application files are inside `apps/web`
- Server starts with no errors in the console

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 11 Pro
  Available memory (MB): 32681
  Available CPU cores: 16
Binaries:
  Node: 24.4.1
  npm: 11.4.2
  Yarn: N/A
  pnpm: N/A
  bun: 1.2.20
Relevant Packages:
  next: 15.4.2-canary.45 // Latest available version is detected (15.4.2-canary.45).
  eslint-config-next: N/A
  react: 19.1.1
  react-dom: 19.1.1
  typescript: 5.8.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Output

### Which stage(s) are affected? (Select all that apply)

next build (local)

### Additional context

Important notes:

- Regression from https://github.com/vercel/next.js/pull/82164
- Issue persists regardless of `--turbopack` build option
- Issue goes away only if both `turbopack.root` and `outputFileTracingRoot` are provided, or only the latter

### Affected Versions

Canary: `15.4.2-canary.22` and higher
Latest working version: `15.4.2-canary.21`

--- a/packages/next/errors.json
+++ b/packages/next/errors.json
@@ -779,5 +779,6 @@
   "778": "`prerenderAndAbortInSequentialTasksWithStages` should not be called in edge runtime.",
   "779": "Route %s used \"searchParams\" inside \"use cache\". Accessing dynamic request data inside a cache scope is not supported. If you need some search params inside a cached function await \"searchParams\" outside of the cached function and pass only the required search params as arguments to the cached function. See more info here: https://nextjs.org/docs/messages/next-request-in-use-cache",
   "780": "Invariant: failed to find parent dynamic route for notFound route %s",
-  "781": "No LRU node to remove"
+  "781": "No LRU node to remove",
+  "782": "Failed to find the root directory of the project. This is a bug in Next.js."
 }

--- a/packages/next/src/lib/find-root.ts
+++ b/packages/next/src/lib/find-root.ts
@@ -17,9 +17,9 @@ export function findRootLockFile(cwd: string) {
   )
 }
 
-export function findRootDir(cwd: string) {
+export function findRootDir(cwd: string): string {
   const lockFile = findRootLockFile(cwd)
-  if (!lockFile) return undefined
+  if (!lockFile) return cwd
 
   const lockFiles = [lockFile]
   while (true) {

--- a/packages/next/src/server/config.ts
+++ b/packages/next/src/server/config.ts
@@ -746,11 +746,15 @@ function assignDefaults(
     result.deploymentId = process.env.NEXT_DEPLOYMENT_ID
   }
 
+  // If either is set, set the other to the same value for tracing root.
   if (result?.outputFileTracingRoot && !result?.turbopack?.root) {
     dset(result, ['turbopack', 'root'], result.outputFileTracingRoot)
   }
+  if (result?.turbopack?.root && !result?.outputFileTracingRoot) {
+    result.outputFileTracingRoot = result.turbopack.root
+  }
 
-  // use the highest level lockfile as tracing root
+  // If neither is set, use the highest level lockfile as tracing root.
   if (!result?.outputFileTracingRoot && !result?.turbopack?.root) {
     let rootDir = findRootDir(dir)
 

--- a/packages/next/src/server/config.ts
+++ b/packages/next/src/server/config.ts
@@ -746,23 +746,28 @@ function assignDefaults(
     result.deploymentId = process.env.NEXT_DEPLOYMENT_ID
   }
 
-  // If either is set, set the other to the same value for tracing root.
-  if (result?.outputFileTracingRoot && !result?.turbopack?.root) {
-    dset(result, ['turbopack', 'root'], result.outputFileTracingRoot)
-  }
-  if (result?.turbopack?.root && !result?.outputFileTracingRoot) {
-    result.outputFileTracingRoot = result.turbopack.root
+  const tracingRoot = result?.outputFileTracingRoot
+  const turbopackRoot = result?.turbopack?.root
+
+  // If both provided, validate they match. If not, use outputFileTracingRoot.
+  if (tracingRoot && turbopackRoot && tracingRoot !== turbopackRoot) {
+    Log.warn(
+      `Both \`outputFileTracingRoot\` and \`turbopack.root\` are set, but they must have the same value.\n` +
+        `Using \`outputFileTracingRoot\` value: ${tracingRoot}.`
+    )
   }
 
-  // If neither is set, use the highest level lockfile as tracing root.
-  if (!result?.outputFileTracingRoot && !result?.turbopack?.root) {
-    let rootDir = findRootDir(dir)
+  const rootDir = tracingRoot || turbopackRoot || findRootDir(dir)
 
-    if (rootDir) {
-      result.outputFileTracingRoot = rootDir
-      dset(result, ['turbopack', 'root'], rootDir)
-    }
+  if (!rootDir) {
+    throw new Error(
+      'Failed to find the root directory of the project. This is a bug in Next.js.'
+    )
   }
+
+  // Ensure both properties are set to the same value
+  result.outputFileTracingRoot = rootDir
+  dset(result, ['turbopack', 'root'], rootDir)
 
   setHttpClientAndAgentOptions(result || defaultConfig)
 

--- a/test/unit/isolated/config.test.ts
+++ b/test/unit/isolated/config.test.ts
@@ -117,4 +117,56 @@ describe('config', () => {
     )
     expect(config.__test__ext).toBe('ts')
   })
+
+  describe('outputFileTracingRoot and turbopack.root consistency', () => {
+    it('Should set both outputFileTracingRoot and turbopack.root to the same value when only outputFileTracingRoot is provided', async () => {
+      const config = await loadConfig(PHASE_DEVELOPMENT_SERVER, '<rootDir>', {
+        customConfig: {
+          outputFileTracingRoot: '/custom/root',
+        },
+      })
+      expect(config.outputFileTracingRoot).toBe('/custom/root')
+      expect(config.turbopack.root).toBe('/custom/root')
+    })
+
+    it('Should set both outputFileTracingRoot and turbopack.root to the same value when only turbopack.root is provided', async () => {
+      const config = await loadConfig(PHASE_DEVELOPMENT_SERVER, '<rootDir>', {
+        customConfig: {
+          turbopack: { root: '/custom/root' },
+        },
+      })
+      expect(config.outputFileTracingRoot).toBe('/custom/root')
+      expect(config.turbopack.root).toBe('/custom/root')
+    })
+
+    it('Should use outputFileTracingRoot value when both are provided with different values', async () => {
+      const config = await loadConfig(PHASE_DEVELOPMENT_SERVER, '<rootDir>', {
+        customConfig: {
+          outputFileTracingRoot: '/tracing/root',
+          turbopack: { root: '/turbo/root' },
+        },
+      })
+      expect(config.outputFileTracingRoot).toBe('/tracing/root')
+      expect(config.turbopack.root).toBe('/tracing/root')
+    })
+
+    it('Should keep the same value when both are provided with matching values', async () => {
+      const config = await loadConfig(PHASE_DEVELOPMENT_SERVER, '<rootDir>', {
+        customConfig: {
+          outputFileTracingRoot: '/same/root',
+          turbopack: { root: '/same/root' },
+        },
+      })
+      expect(config.outputFileTracingRoot).toBe('/same/root')
+      expect(config.turbopack.root).toBe('/same/root')
+    })
+
+    it('Should set both to findRootDir result when neither is provided', async () => {
+      const config = await loadConfig(PHASE_DEVELOPMENT_SERVER, '<rootDir>', {
+        customConfig: {},
+      })
+      expect(config.outputFileTracingRoot).toBeDefined()
+      expect(config.turbopack.root).toBe(config.outputFileTracingRoot)
+    })
+  })
 })

